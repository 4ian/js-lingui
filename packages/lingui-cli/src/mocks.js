export function mockConsole (testCase, mock = {}) {
  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn()
  }

  global.console = {
    ...defaults,
    ...mock
  }

  try {
    testCase(global.console)
  } catch (e) {
    global.console = originalConsole
    throw e
  }

  global.console = originalConsole
}
