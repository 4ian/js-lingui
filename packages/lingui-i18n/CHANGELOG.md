# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.2.3"></a>
## [1.2.3](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.2.2...lingui-i18n@1.2.3) (2017-09-04)




**Note:** Version bump only for package lingui-i18n

<a name="1.2.2"></a>
## [1.2.2](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.2.2-0...lingui-i18n@1.2.2) (2017-08-28)




**Note:** Version bump only for package lingui-i18n

<a name="1.2.1"></a>
## [1.2.1](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.2.0...lingui-i18n@1.2.1) (2017-08-22)




<a name="1.2.0"></a>
# [1.2.0](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.1.0...lingui-i18n@1.2.0) (2017-08-22)


### Features

* Load language data from country-specific locales ([7216d5c](https://github.com/lingui/js-lingui/commit/7216d5c))




<a name="1.1.0"></a>
# [1.1.0](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.0.1...lingui-i18n@1.1.0) (2017-07-12)


### Features

* Add named export of i18n instance ([1ea5aea](https://github.com/lingui/js-lingui/commit/1ea5aea))




<a name="1.0.1"></a>
## [1.0.1](https://github.com/lingui/js-lingui/compare/lingui-i18n@1.0.0...lingui-i18n@1.0.1) (2017-07-10)
