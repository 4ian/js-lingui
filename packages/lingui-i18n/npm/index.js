if (process.env.NODE_ENV === "production") {
  module.exports = require("./cjs/lingui-i18n.production.min.js")
} else {
  module.exports = require("./cjs/lingui-i18n.development.js")
}
