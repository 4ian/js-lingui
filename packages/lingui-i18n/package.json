{
  "name": "lingui-i18n",
  "version": "0.7.0-3",
  "description": "I18n tools for javascript",
  "main": "lib/index.js",
  "module": "lib/index.es.js",
  "author": {
    "name": "TomÃ¡Å¡ Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "license": "MIT",
  "keywords": [
    "i18n",
    "internationalization"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/lingui/js-lingui.git"
  },
  "bugs": {
    "url": "https://github.com/lingui/js-lingui/issues"
  },
  "engines": {
    "node": ">=4.0"
  },
  "dependencies": {
    "babel-runtime": "^6.23.0",
    "lingui-formats": "^0.4.0-3",
    "messageformat-parser": "^1.0.0"
  },
  "scripts": {
    "prepublish": "../../scripts/prepublish.sh"
  }
}
