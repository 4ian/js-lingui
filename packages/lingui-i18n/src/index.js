// @flow
export { default, setupI18n } from './i18n'
export { unpackCatalog } from './loader'

export type { Catalog, Catalogs, Message, LanguageData, I18n } from './i18n'
