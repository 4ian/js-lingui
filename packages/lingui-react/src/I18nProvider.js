import React from "react"

type I18nProps = {
  messages: {[key: string]: string},
  language: string,
  subscribe: Function,
  unsubscribe: Function
}

type I18nProviderProps = I18nProps & {
  children: any
}

/*
 * I18n - Container for i18n data (language, messages)
 * Allows listeners to subscribe for changes
 */
class I18n {
  subscribers = []

  constructor(props) {
    this.update(props)
  }

  subscribe = (callback) => {
    this.subscribers.push(callback)
  }

  unsubscribe = (callback) => {
    this.subscribers = this.subscribers.filter(cb => cb !== callback)
  }

  update = (props) => {
    if (!props) return

    const filteredProps = ['messages', 'language'].reduce((acc, key) => {
      if (props[key]) acc[key] = props[key]
      return acc
    }, {})

    if (!Object.keys(filteredProps).length) return

    this.props = { ...this.props, ...filteredProps }
    this.subscribers.forEach(f => f())
  }

  get messages() {
    return this.props.messages
  }

  get language() {
    return this.props.language
  }
}

class I18nProvider extends React.Component {
  props: I18nProviderProps

  constructor(props) {
    super(props)
    this.i18n = new I18n(props)
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.language !== prevProps.language ||
      this.props.messages !== prevProps.messages
    ) {
      this.i18n.update(this.props)
    }
  }

  getChildContext() {
    return {
      i18n: this.i18n
    }
  }

  render() {
    const { children } = this.props
    return children.length > 1 ? <div>{children}</div> : children
  }
}

I18nProvider.childContextTypes = {
  i18n: React.PropTypes.object
}

export default I18nProvider
export { I18n }
export type { I18nProps }
