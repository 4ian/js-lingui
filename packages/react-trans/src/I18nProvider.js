import React from "react"

type I18nProps = {
  messages: {[key: string]: string},
  language: string,
  subscribe: Function,
  unsubscribe: Function
}

type I18nProviderProps = I18nProps & {
  children: any
}

class I18nProvider extends React.Component {
  props: I18nProviderProps

  subscribers = []

  subscribe = (callback) => {
    this.subscribers.push(callback)
  }

  unsubscribe = (callback) => {
    this.subscribers = this.subscribers.filter(cb => cb !== callback)
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.language !== prevProps.language ||
      this.props.messages !== prevProps.messages
    ) {
      this.subscribers.forEach(f => f())
    }
  }

  getChildContext() {
    const { messages, language } = this.props

    return {
      i18n: {
        messages, language,
        subscribe: this.subscribe,
        unsubscribe: this.unsubscribe
      }
    }
  }

  render() {
    const { children } = this.props
    return children.length > 1 ? <div>{children}</div> : children
  }
}

I18nProvider.childContextTypes = {
  i18n: React.PropTypes.object
}

export default I18nProvider
export type { I18nProps }
