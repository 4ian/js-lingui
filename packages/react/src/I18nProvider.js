// @flow
import * as React from "react"
import type { I18n } from "@lingui/core"

export const {
  Provider: I18nCoreProvider,
  Consumer: I18nCoreConsumer
} = React.createContext({})
export const {
  Provider: I18nDefaultRenderProvider,
  Consumer: I18nDefaultRenderConsumer
} = React.createContext(null)

export type I18nProviderProps = {
  i18n: I18n,
  children: any,
  defaultRender?: any
}

export type I18nProviderState = {
  i18n: I18n
}

export default class I18nProvider extends React.Component<
  I18nProviderProps,
  I18nProviderState
> {
  unsubscribe: Function

  constructor(props: I18nProviderProps) {
    super(props)
    this.state = {
      i18n: this.props.i18n
    }
  }

  componentDidMount() {
    this.unsubscribe = this.props.i18n.onActivate(this.localeChanged)
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  localeChanged = () => {
    const i18n = Object.assign(
      Object.create(Object.getPrototypeOf(this.props.i18n)),
      this.props.i18n
    )
    this.setState({ i18n })
  }

  render() {
    return (
      <I18nDefaultRenderProvider value={this.props.defaultRender}>
        <I18nCoreProvider value={this.state.i18n}>
          {this.props.children}
        </I18nCoreProvider>
      </I18nDefaultRenderProvider>
    )
  }
}
