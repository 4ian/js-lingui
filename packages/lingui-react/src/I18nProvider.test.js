// @flow
import React from 'react'
import { shallow, mount } from 'enzyme'

import { I18nProvider } from '.'
import { LinguiPublisher } from './I18nProvider'
import { mockConsole } from './mocks'
import linguiDev from './dev'
import { setupI18n } from 'lingui-i18n'

describe('I18nProvider', function () {
  const props = {
    messages: {
      cs: {
        'All human beings are born free and equal in dignity and rights.': 'Všichni lidé rodí se svobodní a sobě rovní co do důstojnosti a práv.'
      },
      en: {}
    },
    language: 'cs'
  }

  it('should provide context with i18n data', function () {
    const component = shallow(<I18nProvider {...props}>
      <div />
    </I18nProvider>).instance()
    const linguiPublisher = component.getChildContext()['linguiPublisher']
    expect(linguiPublisher.i18n.messages).toEqual(props.messages[props.language])
    expect(linguiPublisher.i18n.language).toEqual(props.language)
    expect(linguiPublisher.subscribe).toBeInstanceOf(Function)
    expect(linguiPublisher.unsubscribe).toBeInstanceOf(Function)
    expect(linguiPublisher.update).toBeInstanceOf(Function)
  })

  it('should throw an error on incorrect language', function () {
    mockConsole(console => {
      mount(<I18nProvider language="xyz" messages={{}}><div/></I18nProvider>)
      expect(console.warn).toBeCalledWith('Message catalog for locale "xyz" not loaded.')
    })
  })

  it('should render children', function () {
    const child = <div className="testcase"/>
    expect(shallow(
      <I18nProvider {...props}>{child}</I18nProvider>
    ).contains(child)).toBeTruthy()
  })

  it('should render multiple children wrapped in div', function () {
    expect(shallow(
      <I18nProvider {...props}>
        <span />
        <span />
      </I18nProvider>
    ).find('div').length).toEqual(1)
  })

  it('should notify all subscribers about context change', function () {
    const node = mount(<I18nProvider language="en" messages={{
      en: {},
      cs: {}
    }}>
      <div />
    </I18nProvider>)
    const instance = node.instance()
    const listener = jest.fn()

    // $FlowIgnore - Instance returned from enzyme doesn't have custom attrs
    instance.linguiPublisher.subscribe(listener)
    expect(listener).not.toBeCalled()

    node.setProps({ language: 'en' })
    expect(listener).not.toBeCalled()

    node.setProps({ language: 'cs' })
    expect(listener).toBeCalled()
  })
})

describe('I18nPublisher', function () {
  it('should pass active language and messages to underlying I18n class', function () {
    const linguiPublisher = new LinguiPublisher(setupI18n({
      language: 'en',
      messages: {
        en: { msg: 'hello' },
        fr: { msg: 'salut' }
      },
      development: linguiDev
    }))

    expect(linguiPublisher.i18n.language).toEqual('en')
    expect(linguiPublisher.i18n.messages).toEqual({ msg: 'hello' })

    linguiPublisher.update({ language: 'fr' })
    expect(linguiPublisher.i18n.language).toEqual('fr')
    expect(linguiPublisher.i18n.messages).toEqual({ msg: 'salut' })

    linguiPublisher.update({ messages: { fr: { msg: 'salut!' } } })
    expect(linguiPublisher.i18n.language).toEqual('fr')
    expect(linguiPublisher.i18n.messages).toEqual({ msg: 'salut!' })

    linguiPublisher.update({ languageData: { fr: { plurals: () => 'Function' } } })
    expect(linguiPublisher.i18n.language).toEqual('fr')
    expect(linguiPublisher.i18n.messages).toEqual({ msg: 'salut!' })
    expect(linguiPublisher.i18n.languageData.plurals()).toEqual('Function')
  })

  it('should subscribe/unsubscribe listeners for context changes', function () {
    const linguiPublisher = new LinguiPublisher(setupI18n({
      language: 'en',
      messages: { en: {}, fr: {} }
    }))
    const listener = jest.fn()

    expect(linguiPublisher.getSubscribers()).toEqual([])

    linguiPublisher.subscribe(listener)
    expect(linguiPublisher.getSubscribers()).toEqual([listener])

    linguiPublisher.unsubscribe(listener)
    expect(linguiPublisher.getSubscribers()).toEqual([])
  })

  it('should notify listeners only when relevant data changes', function () {
    const listener = jest.fn()
    const linguiPublisher = new LinguiPublisher(setupI18n({
      language: 'en',
      messages: { en: {}, fr: {} },
      development: linguiDev
    }))
    linguiPublisher.subscribe(listener)

    expect(listener).not.toBeCalled()

    linguiPublisher.update()
    expect(listener).not.toBeCalled()

    linguiPublisher.update({ completelyDifferentProp: 42 })
    expect(listener).not.toBeCalled()

    linguiPublisher.update({ language: 'fr' })
    expect(listener).toBeCalled()
    listener.mockReset()

    linguiPublisher.update({ messages: { en: { id: 'hello' } } })
    expect(listener).toBeCalled()
  })
})
