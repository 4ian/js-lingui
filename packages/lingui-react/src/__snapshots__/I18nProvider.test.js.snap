// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18nProvider should throw an error on incorrect language 1`] = `"Unknown locale \\"xyz\\"."`;
